File: java/src/main/java/com/example/backend/application/UserManageService.java
package com.example.backend.application;

import com.example.backend.application.lock.LockKeys;
import com.example.backend.application.lock.LockService;
import com.example.backend.domain.User;
import com.example.backend.domain.UserRepository;
import com.example.backend.domain.UserUniqueSpecification;
import org.springframework.stereotype.Component;

/**
 * 用户管理 Service
 */
@Component
public class UserManageService {

    private final UserRepository userRepository;
    private final UserUniqueSpecification userUniqueSpecification;
    private final LockService lockService;

    public UserManageService(
            UserRepository userRepository,
            UserUniqueSpecification userUniqueSpecification,
            LockService lockService
    ) {
        this.userRepository = userRepository;
        this.userUniqueSpecification = userUniqueSpecification;
        this.lockService = lockService;
    }

    /**
     * 设置用户管理员状态
     */
    public void setAdmin(String id, boolean isAdmin) {
        User user = this.userRepository.selectByIdReq(id);
        user.setAdmin(isAdmin);
        this.userRepository.update(user);
    }

    /**
     * 创建用户
     */
    public String createUser(String username, String password, String nickname, String mobile) {
        try(AutoCloseable lock = this.lockService.lock(LockKeys.USER)){
            User user = User.createUser(
                    IdGenerator.generateId(), username, password, nickname, mobile, this.userUniqueSpecification
            );
            this.userRepository.insert(user);
            return user.getId();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * 修改用户
     */
    public void modifyUser(String id, String username, String nickname, String mobile) {
        try(AutoCloseable lock = this.lockService.lock(LockKeys.USER)){
            User user = this.userRepository.selectByIdReq(id);
            user.modify(username, nickname, mobile, this.userUniqueSpecification);
            this.userRepository.update(user);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * 删除用户
     */
    public void removeUser(String id) {
        this.userRepository.deleteById(id);
    }
}

File: backend-example-dotnet/src/application/UserManageService.cs
using BackendExample.Domain;

namespace BackendExample.Application
{
    /// <summary>
    /// 用户管理 Service
    /// </summary>
    public class UserManageService
    {
        private readonly UserRepository userRepository;
        private readonly UserUniqueSpecification userUniqueSpecification;
        private readonly LockService lockService;

        public UserManageService(
            UserRepository userRepository,
            UserUniqueSpecification userUniqueSpecification,
            LockService lockService)
        {
            this.userRepository = userRepository;
            this.userUniqueSpecification = userUniqueSpecification;
            this.lockService = lockService;
        }

        /// <summary>
        /// 设置用户管理员状态
        /// </summary>
        public async Task SetAdmin(string id, bool isAdmin)
        {
            User user = await this.userRepository.SelectByIdReq(id);
            user.SetAdmin(isAdmin);
            await this.userRepository.Update(user);
        }

        /// <summary>
        /// 创建用户
        /// </summary>
        public async Task<string> CreateUser(string username, string password, string nickname, string mobile)
        {
            await using(await this.lockService.LockAsync(LockKeys.USER))
            {
                User user = await User.Create(IdGenerator.GenerateId(), username, password, nickname, mobile,
                    this.userUniqueSpecification);
                await this.userRepository.Insert(user);
                return user.Id;
            }
        }

        /// <summary>
        /// 修改用户
        /// </summary>
        public async Task ModifyUser(string id, string username, string nickname, string mobile)
        {
            await using (await this.lockService.LockAsync(LockKeys.USER))
            {
                User user = await this.userRepository.SelectByIdReq(id);
                await user.Modify(username, nickname, mobile, this.userUniqueSpecification);
                await this.userRepository.Update(user);
            }
        }

        /// <summary>
        /// 删除用户
        /// </summary>
        public async Task RemoveUser(string id)
        {
            await this.userRepository.DeleteById(id);
        }
    }
}
